qubrabench/benchmark.py

```diff
@@ -0,0 +1,297 @@
+    quantum_expected_quantum_queries: Optional[float] = None
+
+    def _record_query(self, *, n: int, only_actual: bool):
+        """Record an actual classical query.
+
+        Propagates it to the expected stats if they are already computed.
+        """
+
+        self.classical_actual_queries += n
+        if not only_actual:
+            if self.classical_expected_queries is not None:
+                self.classical_expected_queries += n
+            if self.quantum_expected_classical_queries is not None:
+                self.quantum_expected_classical_queries += n
+
+    def as_benchmarked(self):
+        """Propagate the recorded true queries."""
+
+        return QueryStats(
+            classical_actual_queries=self.classical_actual_queries,
+            classical_expected_queries=(
+                self.classical_expected_queries
+                if self.classical_expected_queries is not None
+                else self.classical_actual_queries
+            ),
+            quantum_expected_classical_queries=(
+                self.quantum_expected_classical_queries
+                if self.quantum_expected_classical_queries is not None
+                else self.classical_actual_queries
+            ),
+            quantum_expected_quantum_queries=(
+                self.quantum_expected_quantum_queries
+                if self.quantum_expected_quantum_queries is not None
+                else 0
+            ),
+        )
+
+    def __add__(self, other: "QueryStats") -> "QueryStats":
+        lhs, rhs = self.as_benchmarked(), other.as_benchmarked()
+        return QueryStats(
+            classical_actual_queries=(
+                lhs.classical_actual_queries + rhs.classical_actual_queries
+            ),
+            classical_expected_queries=(
+                lhs.classical_expected_queries + rhs.classical_expected_queries
+            ),
+            quantum_expected_classical_queries=(
+                lhs.quantum_expected_classical_queries
+                + rhs.quantum_expected_classical_queries
+            ),
+            quantum_expected_quantum_queries=(
+                lhs.quantum_expected_quantum_queries
+                + rhs.quantum_expected_quantum_queries
+            ),
+        )
+
+
+class BenchmarkFrame:
+    """A benchmark stack frame, mapping oracle hashes to their computed stats"""
+
+    stats: dict[int, QueryStats]
+    _track_only_actual: bool
+
+    def __init__(self):
+        self.stats = dict()
+        self._track_only_actual = False
+
+    def get_stats(self, obj: object) -> QueryStats:
+        """Get the statistics of a quantum oracle/data structure"""
+
+        h = self.__get_hash(obj)
+        if h not in self.stats:
+            raise ValueError(f"object {obj} has not been benchmarked!")
+        return self.stats[h]
+
+    def __get_hash(self, obj: object) -> int:
+        """hashing used to store the stats"""
+        return hash(obj)
```

Comment:
As some data types are not hashable, I was greeted with an array of error messages, even though the error that caused it (hashing an object) was miniscule. Maybe we can make the user experience better somehow?

Reply:
Also, we implemented hashing for SatInstances, etc., but users don't know that they have to make their constructs hashable?